### DISCUSSION TITLE
Userland filter function

### DISCUSSION THREAD

**Discussion by Baptouuuu** [vimeo/psalm#8031] [Ideas]
Psalm is able to correctly determine the return type of an array after a call to `array_filter`. If we use a templated _collection_ class  that have a `filter` method (with the same logic as `array_filter`) Psalm is not able to determine the correct template of the container after the call to `filter`.

Example: https://psalm.dev/r/9cab8b9bdc

It would be interesting to have a way to declare this behaviour but I'm not sure how the syntax would look like.

**Comment by orklah** [comment]
array_filter is a know function for which we defined a special behaviour here: https://github.com/vimeo/psalm/blob/9d225086e9294cd311d032f8715e27319883614f/src/Psalm/Internal/Provider/ReturnTypeProvider/ArrayFilterReturnTypeProvider.php#L36

However, the method "filter" of a random object doesn't have this particular behaviour applied by Psalm.

If it's part of a framework, you could consider creating a plugin to recycle this behaviour when you spot the desired method in an object

**Comment by AndrolGenhald** [comment]
Not currently possible, but it would probably be possible if templates and closures were improved. The syntax would be a bit verbose though. See #7944 and [this](https://psalm.dev/r/b048bfcf34) example (not working, but how it probably could work in the future).
